### DISCUSSION TITLE
Implementing PrismaClient in an Angular13 project

### DISCUSSION THREAD

**Discussion by MwickFR** [prisma/prisma#12982] [Q&A]
Hi everyone,
I would really appreciate some help on how to handle prisma ORM in Angular 13.
I followed the step by step guide form Prisma website here : [https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases-typescript-mysql](prisma_start_from_scratch);
My tsconfig.json file has some differences and I don't know if it is important :
- "lib" contains "es2020" and "dom";
- "outDir" contains "./dist/out-tsc" instead;

I do successfully connect to my local, MySQL database so, I think it's not so relevant...

```
PS ***>npx prisma migrate dev --name init
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": MySQL database "projektor" at "localhost:3306"

Applying migration "20220424111208_init"

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20220424111208_init/
    └─ migration.sql

Your database is now in sync with your schema.

✔ Generated Prisma Client (3.12.0 | library) to .\node_modules\@prisma\client in 152ms
```
Then...
```
PS ***>npx prisma db pull
Prisma schema loaded from prisma\schema.prisma
Environment variables loaded from .env
Datasource "db": MySQL database "projektor" at "localhost:3306"

Introspecting based on datasource defined in prisma\schema.prisma …

✔ Introspected 2 models and wrote them into prisma\schema.prisma in 95ms

Run prisma generate to generate Prisma Client.
```
And finally...
```
PS ***>npx prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma

✔ Generated Prisma Client (3.12.0 | library) to .\node_modules\@prisma\client in 242ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
****
import { PrismaClient } from "@prisma/client"
const prisma = new PrismaClient()
****
```
I then try to access the 2 fake users stored inside the database.
So I created a basic TS file with content similar to prisma's tutorial :
```
import { PrismaClient, user } from "@prisma/client";

const prismaClient = new PrismaClient();

async function main() {
  // ... you will write your Prisma Client queries here
  const allUsers = await prismaClient.user.findMany()
  console.log(allUsers)
}
```

Calling the "npx ts-node path_to_the_file.ts" returns me an error.
`(node:11108) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use "node --trace-warnings ..." to show where the warning was created)
    at wrapSafe (node:internal/modules/cjs/loader:1031:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Module.m._compile (****\node_modules\ts-node\src\index.ts:1455:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Object.require.extensions.<computed> [as .ts] (****\node_modules\ts-node\src\index.ts:1458:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at phase4 (****\node_modules\ts-node\src\bin.ts:567:12)`

And, at the same time, I just try log, in my Angular active project, the same action which results in :
`Uncaught Error: PrismaClient is unable to be run in the browser.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    PrismaClient index-browser.js:108
    4066 database-handler.service.ts:4
    Webpack 15
        __webpack_require__
        6541
        __webpack_require__
        5041
        __webpack_require__
        6747
        __webpack_require__
        4431
        __webpack_require__
        __webpack_exec__
        <anonymous>
        O
        <anonymous>
        webpackJsonpCallback
        <anonymous>
index-browser.js:108`
The described line 4 of "database-handler.service.ts" is where I instanciate the PrismaClient, as mentionned upper.

So I don't where to introspect as it is the strict description of the tutorial.

Thanks in advance for your help!
My package.json dependencies :
`"dependencies": {
    "@angular-builders/custom-webpack": "^13.1.0",
    "@angular/animations": "~13.1.0",
    "@angular/cdk": "github:angular/cdk-builds",
    "@angular/common": "~13.1.0",
    "@angular/compiler": "~13.1.0",
    "@angular/core": "~13.1.0",
    "@angular/forms": "~13.1.0",
    "@angular/material": "github:angular/material2-builds",
    "@angular/platform-browser": "~13.1.0",
    "@angular/platform-browser-dynamic": "~13.1.0",
    "@angular/router": "~13.1.0",
    "@types/debug": "^4.1.7",
    "@types/sequelize": "^4.28.11",
    "@types/validator": "^13.7.2",
    "assert": "^2.0.0",
    "bootstrap": "^5.1.3",
    "crypto": "^1.0.1",
    "crypto-js": "^4.1.1",
    "fs": "^0.0.1-security",
    "graceful-fs": "^4.2.9",
    "hammerjs": "^2.0.8",
    "materialize-css": "^1.0.0-rc.2",
    "moment": "^2.29.3",
    "mysql2": "^2.3.3",
    "node": "^16.13.2",
    "path": "^0.12.7",
    "pg-hstore": "^2.3.4",
    "reflect-metadata": "^0.1.13",
    "request": "^2.88.2",
    "rxjs": "~7.4.0",
    "semantic-release": "^19.0.2",
    "sequelize-typescript": "^2.1.3",
    "tslib": "^2.3.0",
    "util": "^0.12.4",
    "zone.js": "~0.11.4"
  }`

Aymeric.

**Comment by ruheni** [comment]
Hi 👋🏽

Prisma runs on the server-side, and it seems you're running it on the browser/ client-side. I would recommend using [Angular Universal](https://angular.io/guide/universal) to write Prisma queries or build a separate API in which you can use Prisma to write your queries. You can use one of the [Prisma Examples](https://github.com/prisma/prisma-examples) to get you started building an API quickly/ fullstack example or take a look at the [production ready examples](https://github.com/prisma/prisma-examples#real-world--production-ready-example-projects-with-prisma).

**Comment by MwickFR** [comment]
Thanks for your answer, Ruheni, but I don't get the point.
I do understand that Prisma runs server-side, nevertheless, it seems that the couple Angular-Prisma is not enough to do the complete job.
The Prisma Examples page link presents more than 10 real life examples, but none of them on an Angular basis !?!
Moreover, a large part of the examples requires Express (and Angular Universal too...) which leads, since Webpack 5, to a cascade of compiling errors.
```
BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

on-finished:196:11-33 - Warning: Module not found: Error: Can't resolve 'async_hooks' in '***\node_modules\body-parser\node_modules\on-finished'
./node_modules/express/lib/view.js:81:13-25 - Warning: Critical dependency: the request of a dependency is an expression
./node_modules/express/node_modules/on-finished/index.js:196:11-33 - Warning: Module not found: Error: Can't resolve 'async_hooks' in '***\node_modules\express\node_modules\on-finished'
./node_modules/raw-body/index.js:284:11-33 - Warning: Module not found: Error: Can't resolve 'async_hooks' in '***\node_modules\raw-body'
on-finished:196:11-33 - Warning: Module not found: Error: Can't resolve 'async_hooks' in '***\node_modules\send\node_modules\on-finished'
./node_modules/body-parser/lib/read.js:20:11-26 - Error: Module not found: Error: Can't resolve 'zlib' in '***\node_modules\body-parser\lib'
./node_modules/body-parser/lib/types/urlencoded.js:228:12-34 - Error: Module not found: Error: Can't resolve 'querystring' in '***\node_modules\body-parser\lib\types'
./node_modules/destroy/index.js:17:13-30 - Error: Module not found: Error: Can't resolve 'stream' in '***\node_modules\destroy'
...
```
_(I had the same issues when I tried to use **Sequelize**, which pushed me to try another client...)_

So could it be possible to have an initial list of the required packages to link an **Angular 13** project to a **local MySQL Database** on using **Prisma** ? _(What is your classical config and/or what package(s) you recommend)_

Can I cross/mix the config files of Prisma and Angular Universal to have an easy-to-read project ? ('_.env_' file inside the '_server.ts_' file for example)

If you could point me somewhere to a simple working Angular 10+ project _(presenting a pure basic User and Post MySQL Prisma schema)_ displaying an initial request of the user's list on a html page, I think that I could better understand my mistakes and also the compiling errors I'm confronted with.

Sorry to bother and, again, thanks for your help.
Regards.

**Reply by janpio** [reply]
Angular is usually used for a Client side application.
Those do not directly talk to databases, but to a server first - then then talks to the database.
That server can be implemented with express for examples.

Angular <-> Express (which internally uses Prisma or Sequelize to talk to the database)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
